crates/bevy_ecs/src/schedule_v3/state.rs

```diff
@@ -0,0 +1,50 @@
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::mem;
+
+use crate as bevy_ecs;
+use crate::change_detection::Mut;
+use crate::schedule_v3::{ScheduleLabel, SystemSet, WorldExt};
+use crate::system::Resource;
+use crate::world::World;
+
+/// Types that can define states in a finite-state machine.
+pub trait Statelike: 'static + Send + Sync + Clone + PartialEq + Eq + Hash + Debug {}
+impl<T> Statelike for T where T: 'static + Send + Sync + Clone + PartialEq + Eq + Hash + Debug {}
+
+/// TBD
+#[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct OnEnter<S: Statelike>(pub S);
+
+/// TBD
+#[derive(ScheduleLabel, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct OnExit<S: Statelike>(pub S);
+
+/// TBD
+#[derive(SystemSet, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct OnUpdate<S: Statelike>(pub S);
+
+/// A finite-state machine whose transitions have associated schedules
+/// ([`OnEnter(state)`] and [`OnExit(state)`]).
+///
+/// A state transition can be queued through the [`Transition<S>`] resource, and it will
```

Comment:
```suggestion
/// The current `State<S>` value can be accessed as a resource. 
/// To change the state however,  queue up a state transition through the [`Transition<S>`] resource, and it will
```